draft-ietf-jsonpath-base.md

```diff
@@ -1255,13 +1255,16 @@ Children of an array appear in array order in the resultant nodelist.
 {: unnumbered}
 
 A path by itself in a Logical context is an existence test which yields true if the path selects at least one node and yields false if the path does not select any nodes.
+A function expression with result type `SingleNodeType` or `NodesType` in a Logical context is an existence test
+which yields true if the function returns a non-empty nodelist and yields false if the function returns an empty node-list.
 
 Existence tests differ from comparisons in that:
 
 * they work with arbitrary relative or absolute paths (not just Singular Paths).
 * they work with paths that select structured values.
+* they work with function expressions with result type `SingleNodeType` or `NodesType` (not just `SingleNodeType`).
 
-To test the value of a node selected by a path, an explicit comparison is necessary.
+To test the value of a node selected by a path or returned by a function expression, an explicit comparison is necessary.
```

Comment:
That function expression might return either

* a value
* a singular-path

right ?

Reply:
In test expressions, valid function expressions return either a LogicalType or a nodelist. They don't return a (JSON) value. The function expressions considered by this sentence are those that return a singular nodelist.

Reply:
It probably helps if you discuss the declared types, which is what the type system constrains.  So you are saying that test expressions take declared types of L, S, and N.  Not V.

Reply:
Yes, that may turn out to be necessary, but currently I am limiting the use of types to the function side of the spec. Rolling out the type system across the whole spec is a much larger enterprise than this PR and would probably benefit from being worked on separately (regardless of whether or not we adopt this PR).

One hesitation I have about rolling out the type system across the whole spec is that the spec might then seem more prescriptive and we'd have to be careful to point out that the types are _abstract_ and need not feature in implementations.